### PR TITLE
Bbrinckm draft01

### PR DESCRIPTION
**Pull request by bbrinckman**
NIPC open-api changes:
- take xml out of the openapi def
- add id (device our group) to register topic (optional for advertisement, mandatory for gatt and conn_events)
- looks multi-attribute at comma separated parameters as a whole, and remove where confusing
- remove group id description in get topic
- get topic supports get all without parameter
- include full topic in topic response
- Add attribute registration
- Update /data/attribute POST/PUT to use a registered attribute
- new GET/DELETE methods leveraging id and attribute in the path
- Move POST /data/attribute/read and write to /data to be used as direct read/write without attribute registration
- Move subscription under /data/attribute and adopt registered attribute approach for all subscription API’s
- Create POST /data/subscription/start & stop for direct subscription without attribute registration
- File is oneOf URL or binary
- reworked bulk API, to support POST / GET / DELETE and attribute registration based approach

### CODE REVIEW

File: nipc-openapi/NIPC_REST.yaml [hunk 1]
```diff
@@ -112,7 +100,7 @@ paths:
       parameters:
         - name: id
           in: query
-          description: device ids that need to be filtered
+          description: device or group id that need to be filtered
           required: false
           explode: true
           schema:
```
Comment (rohitmohan96) [inline nipc-openapi/NIPC_REST.yaml:106]: For comma-separated query parameters, it needs to be defined as such: 
```suggestion
          explode: false
          schema:
            type: array
            items:
              type: string
              format: uuid
              example: 12345678-1234-5678-1234-56789abcdef4
```
This maps to the request: 
`https://localhost/nipc/connectivity/binding?id=12345678-1234-5678-1234-56789abcdef4,12345678-1234-5678-1234-56789abcdef5`

**Review by rohitmohan96 — CHANGES_REQUESTED** [review]

File: nipc-openapi/NIPC_REST.yaml [hunk 2]
```diff
@@ -3310,36 +3128,49 @@ components:
...
+              POST /connectivity/connection:
                 '#/components/schemas/ServiceResponse'
-              /connectivity/connection/delete:
+              DELETE /connectivity/connection:
                 '#/components/schemas/SuccessResponse'
-              /extension/attribute/read:
+              GET /connectivity/connection:
+                '#/components/schemas/SuccessResponse'
+              POST /data/attribute:
+                '#/components/schemas/RegisteredAttributeValue'
```
[the remark this answers is not in the document]
Comment (bbrinckman) [reply nipc-openapi/NIPC_REST.yaml:3157]: fixed typo

File: nipc-openapi/NIPC_REST.yaml [hunk 3]
```diff
@@ -628,36 +555,32 @@ paths:
           content:
             application/json:
               schema:
-                $ref: '#/components/schemas/FailureResponse'
-                
-  /connectivity/services/id/{id}:            
-    get:
+                $ref: '#/components/schemas/FailureResponse' 
+
+  /connectivity/services/selective:
```
[the remark this answers is not in the document]
Comment (bbrinckman) [reply nipc-openapi/NIPC_REST.yaml:560]: it makes it consistent with other API's, if we do POST on an attribute, where it should really be a GET, we name the operation in the API path. If we would not do it here, we would be taking a different approach than with other API paths

File: nipc-openapi/NIPC_REST.yaml [hunk 4]
```diff
@@ -628,36 +555,32 @@ paths:
...
+    post:
```
[the remark this answers is not in the document]
Comment (bbrinckman) [reply nipc-openapi/NIPC_REST.yaml:561]: operationID is not an attibute we have used anywhere. Pls clarify

File: nipc-openapi/NIPC_REST.yaml [hunk 5]
```diff
@@ -589,34 +516,34 @@ paths:
...
       tags:
         - connectivity
-      summary: Discover services on a device
-      description: Discover services on a device
-      operationId: ServiceDiscovey
+      summary: Get all services by id for a device
+      description: |-
+        Get all services by id for a connected device, success
+        when service discovery succeeds failure, failure when 
+        Service discovery fails. This updates cache for cached
```
[the remark this answers is not in the document]
Comment (bbrinckman) [reply nipc-openapi/NIPC_REST.yaml:528]: Explicit Service discovery should always trigger an update of hte GATT cache

File: nipc-openapi/NIPC_REST.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (rohitmohan96) [reply nipc-openapi/NIPC_REST.yaml:561]: https://github.com/iot-onboarding/non-ip-iot-control/blob/462bbd267989f5d360465ab86f40d50494d5bdc4/nipc-openapi/NIPC_REST.yaml#L64
It is used in other places but is missing here.

File: nipc-openapi/NIPC_REST.yaml (hunk 1 above)
Comment (bbrinckman) [reply nipc-openapi/NIPC_REST.yaml:106]: updated, pls check

Comment (rohitmohan96) [reply nipc-openapi/NIPC_REST.yaml:106]: Have you pushed the change? I Don't see it.

File: nipc-openapi/NIPC_REST.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (bbrinckman) [reply nipc-openapi/NIPC_REST.yaml:561]: added

File: nipc-openapi/NIPC_REST.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (bbrinckman) [reply nipc-openapi/NIPC_REST.yaml:560]: I am not married to selective, discover is fine

File: nipc-openapi/NIPC_REST.yaml [hunk 6]
```diff
@@ -2843,40 +2656,33 @@ components:
...
             time between reads in seconds (default 1, max 60)
           type: integer
-      xml:
-        name: AttributeConditional
 
 ## A subscription attribute of an Device
-    Subscription:
+    SubscriptionRaw:
       allOf:
         - $ref: '#/components/schemas/Attribute'
```
[the remark this answers is not in the document]
Comment (bbrinckman) [reply nipc-openapi/NIPC_REST.yaml:2663]: fixed

File: nipc-openapi/NIPC_REST.yaml [hunk 7]
```diff
@@ -112,9 +100,9 @@ paths:
...
         - name: id
           in: query
-          description: device ids that need to be filtered
+          description: device or group id that need to be filtered
           required: false
-          explode: true
+          explode: false
           schema:
             type: string
             format: uuid
```
Comment (rohitmohan96) [inline nipc-openapi/NIPC_REST.yaml:108]: For a comma separated list, the type must be array and not string. 
```suggestion
            type: array
            items:
              type: string
              format: uuid
```

**Review by rohitmohan96 — CHANGES_REQUESTED** [review]

File: nipc-openapi/NIPC_REST.yaml [hunk 8]
```diff
@@ -2945,27 +2748,37 @@ components:
...
           enum:
             - default
-            - timestamped
             - payload
         replay:
           type: string
           example: off
           enum:
             - off #default
             - on
```
Comment (rohitmohan96) [inline nipc-openapi/NIPC_REST.yaml:2757]:
```suggestion
          type: boolean
          example: false
          default: false
```

File: nipc-openapi/NIPC_REST.yaml [hunk 9]
```diff
@@ -3004,37 +2829,53 @@ components:
             operation:
```
Comment (rohitmohan96) [inline nipc-openapi/NIPC_REST.yaml:2829]:
```suggestion
            path:
```

File: nipc-openapi/NIPC_REST.yaml [hunk 10]
```diff
@@ -3004,37 +2829,53 @@ components:
...
-               - /extension/connection/delete
-               - /extension/attribute/read
-               - /extension/attribute/write
+               - /connectivity/connection
+               - /data/attribute
+               - /data/attribute/subscription
                - /extension/attribute/write/file
                - /extension/attribute/write/blob
                - /extension/attribute/read/conditional
+            type:
```
Comment (rohitmohan96) [inline nipc-openapi/NIPC_REST.yaml:2838]:
```suggestion
            method:  
```

**Review by rohitmohan96 — APPROVED** [review]

**Pull request merged**